apisix/discovery/zookeeper/init.lua

```diff
@@ -0,0 +1,209 @@
+    if not services then
+        utils.close_zk_client(client)
+        log.error("get zk root children failed: ", err)
+        return
+    end
+
+    -- 3. Fetch the instances of each service and update the cache
+    local now = ngx.time()
+    for _, service in ipairs(services) do
+        local instances, err = fetch_service_instances(local_conf, service)
+        if instances then
+            instance_cache:set(service, nil, {
+                nodes = instances,
+                expire_time = now + local_conf.cache_ttl
+            })
+        else
+            log.error("fetch service instances failed: ", service, " err: ", err)
+        end
+    end
+
+    -- 4. Close connects
+    utils.close_zk_client(client)
+end
+
+function _M.nodes(service_name)
+    if not service_name or service_name == "" then
+        log.error("service name is empty")
+        return nil
+    end
+
+    -- 1. Get from cache
+    local cache = instance_cache:get(service_name)
+    local now = ngx.time()
+
+    -- 2. If the cache is missed or expired, actively pull (the data)）
+    if not cache or cache.expire_time < now then
+        log.debug("cache miss or expired, fetch from zk: ", service_name)
+        local instances, err = fetch_service_instances(local_conf, service_name)
+        if not instances then
+            log.error("fetch instances failed: ", service_name, " err: ", err)
+            -- Fallback: Return the old cache (if available)）
+            if cache then
+                return cache.nodes
+            end
+            return nil
+        end
+
+        -- Update the cache
+        cache = {
+            nodes = instances,
+            expire_time = now + local_conf.cache_ttl
+        }
+        instance_cache:set(service_name, nil, cache)
+    end
+
+    return cache.nodes
+end
+
+function _M.check_schema(conf)
+    return schema.check(conf)
+end
+
+function _M.init_worker()
+    -- Load configuration
+    local core_config = core.config.local_conf
+    local_conf = core_config.discovery and core_config.discovery.zookeeper or {}
+    local ok, err = schema.check(local_conf)
+    if not ok then
+        log.error("invalid zookeeper discovery config: ", err)
+        return
+    end
+
+    -- The default values
+    local_conf.connect_string = local_conf.connect_string or "127.0.0.1:2181"
+    local_conf.fetch_interval = local_conf.fetch_interval or 10
+    local_conf.cache_ttl = local_conf.cache_ttl or 30
+
+    -- Start the timer
+    if not fetch_timer then
+        fetch_timer = ngx.timer.every(local_conf.fetch_interval, fetch_all_services)
+        log.info("zk discovery fetch timer started, interval: ", local_conf.fetch_interval, "s")
+    end
+
+    -- Manually execute a full pull immediately
+    ngx.timer.at(0, fetch_all_services)
+end
+
+function _M.destroy()
+    if fetch_timer then
+        fetch_timer = nil
+    end
+    instance_cache:flush_all()
+    log.info("zookeeper discovery destroyed")
+end
+
```

Comment:
A dump method is also needed to retrieve all services in the current service discovery instance. You can explore other service discovery modules for more information.